File: ports/jagentic-core/src/main/java/org/jagentic/core/Retrieval.java
package org.jagentic.core;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.function.BiFunction;

/**
 * Live hot+cold retrieval — portable analogue of {@code HotVectorIndex} +
 * {@code TwoTierRetriever} + the deterministic hashing embedder. Model-free so the
 * port runs without an embedding model; swap in a real embedder/cold store later.
 */
public final class Retrieval {
  private Retrieval() {}

  /** Deterministic bag-of-words hashing embedder (L2-normalized). */
  public static float[] embed(String text, int dim) {
    float[] v = new float[dim];
    if (text != null) {
      for (String tok : text.toLowerCase().split("[^a-z0-9]+")) {
        if (tok.isEmpty()) continue;
        int h = tok.hashCode();
        v[Math.floorMod(h, dim)] += (h >>> 31) == 0 ? 1.0f : -1.0f;
      }
    }
    double n = 0;
    for (float x : v) n += (double) x * x;
    n = Math.sqrt(n);
    if (n > 0) for (int i = 0; i < dim; i++) v[i] /= (float) n;
    return v;
  }

  public static double cosine(float[] a, float[] b) {
    if (a.length != b.length) return -1.0;
    double dot = 0, na = 0, nb = 0;
    for (int i = 0; i < a.length; i++) {
      dot += (double) a[i] * b[i];
      na += (double) a[i] * a[i];
      nb += (double) b[i] * b[i];
    }
    return (na == 0 || nb == 0) ? 0 : dot / (Math.sqrt(na) * Math.sqrt(nb));
  }

  /** A scored hit. */
  public record Scored(String id, double score, String text) {}

  /** Recency-bounded vector index (the hot tier). */
  public interface HotVectorIndex {
    void upsert(String id, float[] embedding, String text);

    List<Scored> search(float[] query, int k);

    int size();
  }

  /** Capacity-bounded brute-force cosine window. */
  public static final class InMemoryHotVectorIndex implements HotVectorIndex {
    private record Entry(float[] vec, String text) {}

    private final int max;
    private final LinkedHashMap<String, Entry> entries =
        new LinkedHashMap<>(16, 0.75f, false) {
          @Override
          protected boolean removeEldestEntry(Map.Entry<String, Entry> e) {
            return size() > max;
          }
        };

    public InMemoryHotVectorIndex() {
      this(2000);
    }

    public InMemoryHotVectorIndex(int max) {
      this.max = Math.max(1, max);
    }

    @Override
    public synchronized void upsert(String id, float[] embedding, String text) {
      entries.remove(id);
      entries.put(id, new Entry(embedding, text));
    }

    @Override
    public synchronized List<Scored> search(float[] query, int k) {
      List<Scored> all = new ArrayList<>(entries.size());
      for (Map.Entry<String, Entry> e : entries.entrySet()) {
        all.add(new Scored(e.getKey(), cosine(query, e.getValue().vec()), e.getValue().text()));
      }
      all.sort((x, y) -> Double.compare(y.score(), x.score()));
      return all.subList(0, Math.min(Math.max(1, k), all.size()));
    }

    @Override
    public synchronized int size() {
      return entries.size();
    }
  }

  /** The durable tier as a search seam: (query, k) -> hits. */
  public interface ColdSearch extends BiFunction<float[], Integer, List<Scored>> {}

  /** Merge hot + cold, de-dup by id keeping the higher score, return global top-k. */
  public static final class TwoTierRetriever {
    private final HotVectorIndex hot;
    private final ColdSearch cold;
    private final int hotK;
    private final int coldK;

    public TwoTierRetriever(HotVectorIndex hot, ColdSearch cold, int hotK, int coldK) {
      this.hot = hot;
      this.cold = cold;
      this.hotK = Math.max(1, hotK);
      this.coldK = Math.max(1, coldK);
    }

    public List<Scored> retrieve(float[] query, int k) {
      Map<String, Scored> best = new LinkedHashMap<>();
      if (hot != null) {
        try {
          for (Scored s : hot.search(query, hotK)) merge(best, s);
        } catch (RuntimeException ignored) {
          // degrade to cold-only
        }
      }
      if (cold != null) {
        try {
          List<Scored> c = cold.apply(query, coldK);
          if (c != null) for (Scored s : c) merge(best, s);
        } catch (RuntimeException ignored) {
          // degrade to hot-only
        }
      }
      List<Scored> out = new ArrayList<>(best.values());
      out.sort((x, y) -> Double.compare(y.score(), x.score()));
      return out.subList(0, Math.min(Math.max(1, k), out.size()));
    }

    private static void merge(Map<String, Scored> best, Scored s) {
      Scored prior = best.get(s.id());
      if (prior == null || s.score() > prior.score()) best.put(s.id(), s);
    }
  }
}

File: ports/pyagentic/pyagentic/retrieval.py
"""Live hot+cold retrieval — the portable analogue of ``HotVectorIndex`` +
``TwoTierRetriever`` + the deterministic ``HashingEmbedder`` from the Quarkus
``/rag`` proxy.

Dependency-free so the port runs and is tested without a model: the hashing
embedder is deterministic (bag-of-words → L2-normalized), the hot tier is an
exact brute-force cosine window, and the two-tier retriever merges hot + cold,
de-duping by id (hot wins). Swap the embedder for a real model (DJL/OpenAI) and
the cold tier for pgvector/Qdrant without touching the merge logic.
"""

from __future__ import annotations

import math
import re
from collections import OrderedDict
from dataclasses import dataclass
from threading import RLock
from typing import Callable, Dict, List, Optional, Protocol, Tuple

_TOKEN = re.compile(r"[a-z0-9]+")


def hashing_embedder(dim: int = 64) -> Callable[[str], List[float]]:
    """Deterministic bag-of-words hashing embedder. Documents sharing vocabulary
    land near each other in cosine space — enough to exercise retrieval end-to-end
    without an external model."""

    def embed(text: str) -> List[float]:
        v = [0.0] * dim
        if text:
            for tok in _TOKEN.findall(text.lower()):
                h = hash(tok)
                v[h % dim] += 1.0 if (h >> 63) == 0 else -1.0
        norm = math.sqrt(sum(x * x for x in v))
        return [x / norm for x in v] if norm > 0 else v

    return embed


def cosine(a: List[float], b: List[float]) -> float:
    if len(a) != len(b):
        return -1.0
    dot = sum(x * y for x, y in zip(a, b))
    na = math.sqrt(sum(x * x for x in a))
    nb = math.sqrt(sum(y * y for y in b))
    return dot / (na * nb) if na and nb else 0.0


@dataclass(frozen=True)
class Scored:
    id: str
    score: float
    text: str


class HotVectorIndex(Protocol):
    def upsert(self, doc_id: str, embedding: List[float], text: str) -> None: ...

    def search(self, query: List[float], k: int) -> List[Scored]: ...

    def size(self) -> int: ...


class InMemoryHotVectorIndex:
    """Capacity-bounded (LRU) brute-force cosine window — the recent/hot tier."""

    def __init__(self, max_entries: int = 2000) -> None:
        self._max = max(1, max_entries)
        self._entries: "OrderedDict[str, Tuple[List[float], str]]" = OrderedDict()
        self._lock = RLock()

    def upsert(self, doc_id: str, embedding: List[float], text: str) -> None:
        with self._lock:
            self._entries.pop(doc_id, None)
            self._entries[doc_id] = (embedding, text)
            while len(self._entries) > self._max:
                self._entries.popitem(last=False)

    def search(self, query: List[float], k: int) -> List[Scored]:
        with self._lock:
            scored = [Scored(i, cosine(query, e), t) for i, (e, t) in self._entries.items()]
        scored.sort(key=lambda s: s.score, reverse=True)
        return scored[: max(1, k)]

    def size(self) -> int:
        with self._lock:
            return len(self._entries)


# A cold tier is just a search function: (query_vec, k) -> List[Scored].
ColdSearch = Callable[[List[float], int], List[Scored]]


class TwoTierRetriever:
    """Merge hot + cold, de-dup by id keeping the higher score (hot supersedes a
    stale cold copy), return global top-k. Tolerates either tier failing."""

    def __init__(self, hot: Optional[HotVectorIndex], cold: Optional[ColdSearch], hot_k: int = 8, cold_k: int = 8):
        self.hot = hot
        self.cold = cold
        self.hot_k = max(1, hot_k)
        self.cold_k = max(1, cold_k)

    def retrieve(self, query: List[float], k: int) -> List[Scored]:
        best: Dict[str, Scored] = {}
        if self.hot is not None:
            try:
                for s in self.hot.search(query, self.hot_k):
                    self._merge(best, s)
            except Exception:
                pass
        if self.cold is not None:
            try:
                for s in self.cold(query, self.cold_k):
                    self._merge(best, s)
            except Exception:
                pass
        out = sorted(best.values(), key=lambda s: s.score, reverse=True)
        return out[: max(1, k)]

    @staticmethod
    def _merge(best: Dict[str, Scored], s: Scored) -> None:
        prior = best.get(s.id)
        if prior is None or s.score > prior.score:
            best[s.id] = s
